### PR TITLE
Structured metadata

### LINKED ISSUE #189: Update sweepers to support new structured metadata
### Are you sure this is not a new requirement or bug?

Yes

### 💡 Description

Update sweepers to support new structured metadata.

Related to
- https://github.com/NASA-PDS/harvest/issues/244
- https://github.com/NASA-PDS/registry-api/issues/632

### PR DESCRIPTION
**Pull request by alexdunnjpl**
## 🗒️ Summary
Updates sweepers to use new metadata structure, where flat `/`-separated keys are replaced by opensearch nested paths.

Disables repairkit sweeper as this is understood not to be a necessary part of sweepers anymore - this will be removed once this is confidently confirmed.

Further update will be necessary for managed-opensearch migration, but that's downstream.

### 🤖 AI Assistance Disclosure
- [ ] No AI assistance used
- [ ] AI used for light assistance (e.g., suggestions, refactoring, documentation help, minor edits)
- [ ] AI used for moderate content generation (AI generated some code or logic, but the developer authored or heavily revised the majority)
- [X] AI generated substantial portions of this code

Estimated % of code influenced by AI: 90%

## ⚙️ Test Data and/or Report
Tests pass

## ♻️ Related Issues
Resolves #189 

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Documentation and PR Content
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue
- [ ] **PR Title:** The PR title is "user-friendly" clearly identifying what is being fixed or the new feature being added, that if you saw it in the Release Notes for a tool, you would be able to get the gist of what was done.

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Maintenance
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

### CODE REVIEW

File: src/pds/registrysweepers/utils/misc.py [hunk 1]
```diff
@@ -210,3 +210,51 @@ def limit_log_length(log_msg: str, max_str_length: int = 5012) -> str:
...
+    return has_nested_attr(child_obj, remaining)
+
+
+def get_nested_attr(source: Dict[str, Any], path: str, default: Any = None) -> Any:
+    """
+    Get a value from a (possibly nested) dict using a /-separated path.
+    Single-component paths resolve directly; multi-component paths traverse nested dicts.
+    Returns default if any component in the path is absent or not a dict.
+    """
+    parent, _, remaining = path.partition("/")
```
Comment (nutjob4life) [inline src/pds/registrysweepers/utils/misc.py:235]: This `get_nested_attr` function derives "search-after" values with
```python
get_nested_attr(hit["_source"], field)
```
`get_nested_attr` splits only on `/`, yet we've changed the sort fields to use a "dot" notation; i.e. from

    ops:Harvest_Info/ops:harvest_date_time

to

    ops:Harvest_Info.ops:harvest_date_time

So `get_nested_attr` will be looking for the literal key
```python
"ops:Harvest_Info.ops:harvest_date_time"
```
and not doing
```python
_source["ops:Harvest_Info"]["ops:harvest_date_time"]
```
This'll likely just yield a single-item list results like `[None]`, right?

Affected users:
- `utils/db/__init__.py`
- `legacy_registry_sync/opensearch_loaded_product.py`
- `reindexer/main.py`

Maybe it's possible to pass a slash-path to `get_nessted_attr` separate from OpenSearch sort parths, or teach the helper to resolve both slash and dot notation?

File: src/pds/registrysweepers/reindexer/main.py [hunk 2]
```diff
@@ -45,7 +47,7 @@ def get_docs_query(filter_to_harvested_before: datetime):
                 "must_not": [{"exists": {"field": REINDEXER_FLAG_METADATA_KEY}}],
                 "must": {
                     "range": {
-                        "ops:Harvest_Info/ops:harvest_date_time": {
+                        "ops:Harvest_Info.ops:harvest_date_time": {
```
Comment (nutjob4life) [inline src/pds/registrysweepers/reindexer/main.py:50]: Should `REINDEXER_FLAG_METADATA_KEY` also change to use the new "dot" notation? It's still set to

    ops:Provenance/ops:reindexed_at

File: src/pds/registrysweepers/utils/misc.py [hunk 3]
```diff
@@ -210,3 +210,51 @@ def limit_log_length(log_msg: str, max_str_length: int = 5012) -> str:
...
+
+
+def build_nested_update(flat_updates: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Convert a flat dict of /-separated path→value pairs into a nested dict
+    suitable for use as the content of an OpenSearch partial update.
+    Paths that share a common parent are merged under that parent key.
+    Flat (no-/) paths are included as-is.
+    """
+    result: Dict[str, Any] = {}
```
Comment (nutjob4life) [inline src/pds/registrysweepers/utils/misc.py:251]: This `build_nested_update` function doesn't recurse like `get_nested_attr`/`has_nested_attr`:
```python
parent, _, child = path.partition("/")
result[parent][child] = value
```
So
```python
"a/b/c": 1
```
becomes
```python
{"a": {"b/c": 1}}
```
If all metadata paths are two-level, that's fine, but let's document this, since the docstring says `/-separated path→value pairs into a nested dict` which sort of suggests arbitrarily deep leveling. 

PS: Extra bonus points from using _an actual arrow_ in your docstring! 🎉

File: src/pds/registrysweepers/reindexer/main.py [hunk 4]
```diff
@@ -192,12 +194,9 @@ def accumulate_missing_mappings(
...
 
-                try:
-                    problematic_harvest_versions.update(doc["_source"]["ops:Harvest_Info/ops:harvest_version"])
-                except KeyError as err:
-                    # Noisy log temporarily disabled but may be re-enabled at jpadams' discretion
-                    # log.warning(limit_log_length(f'Unable to extract harvest version from document {doc["_id"]}: {err}'))
-                    pass
+                harvest_version = get_nested_attr(doc["_source"], "ops:Harvest_Info/ops:harvest_version")
+                if harvest_version is not None:
+                    problematic_harvest_versions.update(harvest_version)
```
Comment (nutjob4life) [inline src/pds/registrysweepers/reindexer/main.py:199]: If `harvest_version` is ever a string, this'll add each character of the string. Maybe instead:
```python
problematic_harvest_versions.update(coerce_list_type(harvest_version))
```

File: src/pds/registrysweepers/reindexer/main.py [hunk 5]
```diff
@@ -75,7 +77,7 @@ def fetch_dd_field_types(client: OpenSearch) -> Dict[str, str]:
 
 def get_mapping_field_types_by_field_name(client: OpenSearch, index_name: str) -> Dict[str, str]:
     return {
-        k: v["type"] for k, v in client.indices.get_mapping(index_name)[index_name]["mappings"]["properties"].items()  # type: ignore
+        k: v.get("type") for k, v in client.indices.get_mapping(index_name)[index_name]["mappings"]["properties"].items()  # type: ignore
```
Comment (nutjob4life) [inline src/pds/registrysweepers/reindexer/main.py:80]: Changing to `get` means you can now receive `None` so the type hint on line 78 should become
```python
Dict[str, str | None]
```

**Review by nutjob4life — CHANGES_REQUESTED** [review]
This is a great step in the right direction, with better provenance and less reindexing. I left a few comments left within. Ignore anything that doesn't make sense.

One other idea: the repeated slash/dot conversion might benefit from a dedicated utility. Right now the PR has three path representations:
```
ops:Provenance/ops:ancestor_refs     # internal metadata key
ops:Provenance.ops:ancestor_refs     # OpenSearch query/sort field
{"ops:Provenance": {"ops:ancestor_refs": ...}}  # update/source shape
```
Right now the code relies on remembering which one where has dots and which has slashes. Maybe some explicit helpers?
```python
metadata_path_to_source_path()
metadata_path_to_opensearch_field()
build_nested_update()
```
I don't know! Whatever you think.

File: src/pds/registrysweepers/reindexer/main.py (hunk 4 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/reindexer/main.py:199]: agreed - will fix

File: src/pds/registrysweepers/reindexer/main.py (hunk 5 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/reindexer/main.py:80]: Need to look into the context in a sec, but likely this is intended, to throw (i.e. terminate with error) in the event that a key doesn't exist where it absolutely should.

File: src/pds/registrysweepers/utils/misc.py (hunk 3 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/utils/misc.py:251]: @jordanpadams @tloubrieu-jpl is two-level nesting an intended design choice? i.e. should I document (and put in a guard to catch >2 level keys), or update to support arbitrary-depth nesting?

File: src/pds/registrysweepers/reindexer/main.py (hunk 2 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/reindexer/main.py:50]: This is find as it is translated at the appropriate point (should be and should have been caught by tests if not).

If/when this is refactored, this will be ironed out.

File: src/pds/registrysweepers/utils/misc.py (hunk 1 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/utils/misc.py:235]: Good catch - mea culpa.

**Comment by alexdunnjpl** [comment]
@nutjob4life for context, there was an initial desire way back when to avoid the "clutter" of wrapping every one of a million strings in constructors and method calls, and it _wasn't_ so fraught at that point.

I agree that now, it is showing itself to be brittle enough to need fixing.

@jordanpadams @tloubrieu-jpl are you happy for me to spend the time to refactor it now, or is there urgent pressure to get a working version deployed before resolving this?

File: src/pds/registrysweepers/utils/misc.py (hunk 3 above)
Comment (jordanpadams) [reply src/pds/registrysweepers/utils/misc.py:251]: @alexdunnjpl if I am understanding the question correctly, XML nesting can be any number of nested levels.

File: src/pds/registrysweepers/reindexer/main.py (hunk 5 above)
Comment (nutjob4life) [reply src/pds/registrysweepers/reindexer/main.py:80]: > Need to look into the context in a sec, but likely this is intended, to throw (i.e. terminate with error) in the event that a key doesn't exist where it absolutely should.

In that case you probably want to change it back to
```python
k: v["type"] for k, v in client.indices.get_mapping(index_name)[index_name]["mappings"]["properties"].items()  # type: ignore
```
The indexing of `v["type"]` will throw if it's not there, and then the return type hint of the function, namely,
```python
def get_mapping_field_types_by_field_name(…) -> Dict[str, str]:
```
should stay as is.

If you stay with the PR's newer `v.get("type")`, then it _won't throw_ if it's missing but instead give a `None`, and so the type hint should become:
```python
def get_mapping_field_types_by_field_name(…) -> Dict[str, str | None]:
```

**Comment by alexdunnjpl** [comment]
Replaced by https://github.com/NASA-PDS/registry-sweepers/pull/242
